Add Create, Control, Observe introduction guides (#370)
onkernel/docs: introduction/control.mdx

---
title: "Control"

> description: "Drive the browser with computer use, CDP, or WebDriver BiDi"

Reviewer: Do we feel that the two sections below this flow well enough with the three options presented? It's all only relevant for the computer use choice, but if you flip to cdp or bidi up top, it no longer is relevant.

---

> Kernel browsers expose three control primitives. For agents, we recommend [computer use](/browsers/computer-controls) — the primitives match how computer-use models were trained to drive a computer, and they sidestep the bot-detection surface that CDP introduces.

Reviewer: What do we think of having a forth tab for Playwright Execution API here as well in 2nd position?

        page = context.pages[0]

        await page.goto('https://example.com')
        title = await page.title()
        print(title)
    ```
    </CodeGroup>
  </Tab>
  <Tab title="WebDriver BiDi">
    W3C-standard browser control. Use `webdriver_ws_url` with [Vibium](/integrations/vibium) or any other BiDi client.

    <CodeGroup>
    ```typescript Typescript/Javascript
    import { browser } from 'vibium';

    const bro = await browser.start(kernelBrowser.webdriver_ws_url);
    const page = await bro.page();

    await page.goto('https://example.com');
    const title = await page.title();
    console.log(title);
    ```

    ```python Python
    from vibium.sync_api import browser

    bro = browser.start(kernel_browser.webdriver_ws_url)
    page = bro.page()

    page.goto('https://example.com')
    title = page.title()
    print(title)
    ```
    </CodeGroup>
  </Tab>
</Tabs>

## Why computer use for agents

Kernel's computer controls are built to match how computer-use models were trained — the same primitives the model emits (screenshot, click at coords, type, key, scroll, drag) map 1:1 onto the API. There's no harness translating model output into framework calls.

> - **Native fit.** Screenshot, click, type, key, scroll, drag — the primitives the model already speaks. Kernel uses these same controls in its own [managed auth](/auth/overview) agent.

Reviewer: ```suggestion
- **Native fit.** Screenshot, click, type, key, scroll, drag — the primitives the model already speaks.
```

- **Faster screenshots.** Captures bypass CDP, which removes the largest source of latency in a vision loop.
- **Better against bot detection.** No CDP connection means no CDP fingerprint to leak. Pairs naturally with [stealth mode](/browsers/bot-detection/stealth) and [residential proxies](/proxies/residential).
- **Human-like input.** OS-level events with Bézier-curve mouse paths, variable typing speed, and configurable mistype rate.
- **Not DOM-limited.** Screenshots capture the full VM, so the agent can see and interact with native dialogs, canvas elements, iframes, and PDFs — not just things you can address with a selector.

## Computer use + Playwright execution

> The two things computer controls don't do natively: read the DOM, and take a full-page screenshot. The recommended pattern for agents is **computer controls for interaction, [Playwright execution](/browsers/playwright-execution) as a DOM-reading tool** when the agent needs structured data.

Reviewer: ```suggestion
The one thing computer controls don't do natively: read the DOM. The recommended pattern for agents is computer controls for interaction, [Playwright execution](/browsers/playwright-execution) as a DOM-reading tool when the agent needs structured data.
```

> Playwright execution runs arbitrary Playwright code in a fresh context inside the browser's VM. Your agent can call it as a tool whenever it needs structured DOM data or a full-page capture, then go right back to driving with computer controls. It ships with [Patchright](/browsers/bot-detection/stealth) by default, so DOM-side calls are hardened against bot detection too.

Reviewer: ```suggestion
Playwright execution runs arbitrary Playwright code in a fresh context inside the browser's VM. Your agent can call it as a tool whenever it needs structured DOM data, then go right back to driving with computer controls.
```
